[SPARK-5436] [MLlib] Validate GradientBoostedTrees using runWithValidation

MechCoder opened the pull request:
One can early stop if the decrease in error rate is lesser than a certain tol or if the error increases if the training data is overfit.

This introduces a new method runWithValidation which takes in a pair of RDD's , one for the training data and the other for the validation.

MechCoder committed:
[SPARK-5436] Validate GradientBoostedTrees using runWithValidation

SparkQA commented:
[Test build #27689 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27689/consoleFull) for   PR 4677 at commit [`07c8f12`](https://github.com/apache/spark/commit/07c8f12bc72b11ae780095a73662b5e049dc6e22).
 * This patch merges cleanly.

MechCoder commented:
@jkbradley I just wanted to know if this is in the right direction.

SparkQA commented:
[Test build #27690 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27690/consoleFull) for   PR 4677 at commit [`7534d14`](https://github.com/apache/spark/commit/7534d145d8cf686221647bffeeb2c404dddc575d).
 * This patch merges cleanly.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27689/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27690/
Test FAILed.

shaneknapp commented:
jenkins, test this please

shaneknapp commented:
jenkins, test this please

SparkQA commented:
[Test build #27691 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27691/consoleFull) for   PR 4677 at commit [`7534d14`](https://github.com/apache/spark/commit/7534d145d8cf686221647bffeeb2c404dddc575d).
 * This patch merges cleanly.

jkbradley commented:
My recommendations:
* Name the new run() method something more explicit like runWithValidation().
* Don't add the extra train() methods since users can use run().
* Rename convergenceTol to something like validationTol (since convergenceTol applies to the training set in general).
* Generalize the boost() method rather than duplicating it to avoid code duplication.

We had discussed providing a helper method evaluateEachIteration() in the JIRA, but I'd prefer to have that be a separate JIRA and PR.  Does that sound good?

Thanks!

SparkQA commented:
[Test build #27691 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27691/consoleFull) for   PR 4677 at commit [`7534d14`](https://github.com/apache/spark/commit/7534d145d8cf686221647bffeeb2c404dddc575d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27691/
Test PASSed.

SparkQA commented:
[Test build #27711 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27711/consoleFull) for   PR 4677 at commit [`e008936`](https://github.com/apache/spark/commit/e0089363220f11b53a83bb1afb29af939cbb711f).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27712 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27712/consoleFull) for   PR 4677 at commit [`77549a9`](https://github.com/apache/spark/commit/77549a9b74c510f480df31fad314d3395313812d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27711 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27711/consoleFull) for   PR 4677 at commit [`e008936`](https://github.com/apache/spark/commit/e0089363220f11b53a83bb1afb29af939cbb711f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27711/
Test PASSed.

MechCoder committed:
TST: Add test for classification

SparkQA commented:
[Test build #27712 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27712/consoleFull) for   PR 4677 at commit [`77549a9`](https://github.com/apache/spark/commit/77549a9b74c510f480df31fad314d3395313812d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27712/
Test PASSed.

SparkQA commented:
[Test build #27716 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27716/consoleFull) for   PR 4677 at commit [`3e74372`](https://github.com/apache/spark/commit/3e743723f128f5d87f66dc05a92d3f68e6fd01cb).
 * This patch merges cleanly.

MechCoder commented:
@jkbradley I have fixed up your comments.

Btw, why are there are both a train and a run, which seems to me do the same thing. Is it not better to have one way of doing things.

Also a doubt in the case of the Classification problem. It seems to me for each iteration, the problem is changed explicitly to a Regression problem with labels mapped to {-1, 1}. Is it okay to break when this regression error no longer reduces on the validation data for a classification problem (which seems slightly awkward to me)?

Note that I had to explicitly set it to Regression to make sure that this test passes, (https://github.com/apache/spark/pull/4677/files#diff-d3159b88ae0ed6ff096ff8850ecac26eR207) . Otherwise, the classification error seems to be the same for both with and without validation.

SparkQA commented:
[Test build #27716 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27716/consoleFull) for   PR 4677 at commit [`3e74372`](https://github.com/apache/spark/commit/3e743723f128f5d87f66dc05a92d3f68e6fd01cb).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class Partitioner(object):`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27716/
Test PASSed.

MechCoder committed:
One liner for prevValidateError

SparkQA commented:
[Test build #27720 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27720/consoleFull) for   PR 4677 at commit [`55e5c3b`](https://github.com/apache/spark/commit/55e5c3b22c39ef1f71e35eca0fda06a9080d22d7).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27720 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27720/consoleFull) for   PR 4677 at commit [`55e5c3b`](https://github.com/apache/spark/commit/55e5c3b22c39ef1f71e35eca0fda06a9080d22d7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27720/
Test PASSed.

jkbradley commented:
The train + run combination is a legacy thing, where algorithms were originally called using static methods (train) with a list of arguments.  As algorithms gained more arguments, we moved towards builder patterns, where you build a learner with the parameters/settings you want and then call run.

About Classification, you're getting at a problem inherent in many ML methods: We often learn by optimizing surrogate losses.  I.e., we may care about 0/1 accuracy, not about the logistic loss, but we'll still learn using the logistic loss since the math & optimization work out nicely for it.  I'd recommend:
* For the tests, switching to Regression manually seems reasonable.
* To help users:
  * Add a note in the docs about validation error not changing monotonically.  Recommend to users that they inspect the validation curve manually.
  * Maybe allow negative tolerances.  That would let learning run for a bit longer, but would still stop early when the algorithm really started overfitting.

One more thought: It might be best to compare with the best validation error found so far, rather than the previous iteration's value.  That would prevent slow and steady overfitting.

Thanks for the updates!

MechCoder committed:
Made the following changes
1. Add section to documentation
2. Return corresponding to bestValidationError
3. Allow negative tolerance.

MechCoder commented:
@jkbradley I've fixed up your comments.

a] Negative tol is allowed.
b] It makes sense to return based on the best validationError rather than the previous one (when it breaks) so I've changed that too.
c] I've added a section to ensembles so that this can be easily found.

I think we are ready for a final round of reviews.

SparkQA commented:
[Test build #27765 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27765/consoleFull) for   PR 4677 at commit [`895dcb4`](https://github.com/apache/spark/commit/895dcb44b7362bfe5ffabfdc455fb22f78db020d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27766 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27766/consoleFull) for   PR 4677 at commit [`772ee7e`](https://github.com/apache/spark/commit/772ee7eab53043de321dec29b9b5a50dd1df7d09).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27767 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27767/consoleFull) for   PR 4677 at commit [`fad9b6e`](https://github.com/apache/spark/commit/fad9b6e0ef8d760158410069341e285bebc52c58).
 * This patch merges cleanly.

MechCoder committed:
Move validation while training section under usage tips

SparkQA commented:
[Test build #27768 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27768/consoleFull) for   PR 4677 at commit [`b928a19`](https://github.com/apache/spark/commit/b928a19d4592fb36d2bda0ae600e53d2b240b980).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27765 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27765/consoleFull) for   PR 4677 at commit [`895dcb4`](https://github.com/apache/spark/commit/895dcb44b7362bfe5ffabfdc455fb22f78db020d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27765/
Test PASSed.

SparkQA commented:
[Test build #27766 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27766/consoleFull) for   PR 4677 at commit [`772ee7e`](https://github.com/apache/spark/commit/772ee7eab53043de321dec29b9b5a50dd1df7d09).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27766/
Test PASSed.

SparkQA commented:
[Test build #27767 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27767/consoleFull) for   PR 4677 at commit [`fad9b6e`](https://github.com/apache/spark/commit/fad9b6e0ef8d760158410069341e285bebc52c58).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27767/
Test PASSed.

SparkQA commented:
[Test build #27768 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27768/consoleFull) for   PR 4677 at commit [`b928a19`](https://github.com/apache/spark/commit/b928a19d4592fb36d2bda0ae600e53d2b240b980).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27768/
Test PASSed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala:
```diff
@@ -76,8 +77,42 @@ class GradientBoostedTrees(private val boostingStrategy: BoostingStrategy)
 
+  /**
+   * Method to validate a gradient boosting model
+   * @param input Training dataset: RDD of [[org.apache.spark.mllib.regression.LabeledPoint]].
```
fix param name

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala:
```diff
@@ -76,8 +77,42 @@ class GradientBoostedTrees(private val boostingStrategy: BoostingStrategy)
+   */
+  def runWithValidation(
+      trainInput: RDD[LabeledPoint],
+      validateInput: RDD[LabeledPoint]): GradientBoostedTreesModel = {
```
There is no guarantee that training and validation are following the same distribution. I know it happens in practice but I don't know the theory behind using distribution A for training but another distribution B for validation. It would be better if we say that they should follow the same distribution.

MechCoder replied:
Do you mean to add a comment?

MechCoder committed:
COSMIT

MechCoder commented:
@mengxr Fixed !

SparkQA commented:
[Test build #27775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27775/consoleFull) for   PR 4677 at commit [`77dc5f0`](https://github.com/apache/spark/commit/77dc5f0f0fde0f37a1ba6bb6ffd8f96ccee9beaa).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27775 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27775/consoleFull) for   PR 4677 at commit [`77dc5f0`](https://github.com/apache/spark/commit/77dc5f0f0fde0f37a1ba6bb6ffd8f96ccee9beaa).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27775/
Test FAILed.

SparkQA commented:
[Test build #27776 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27776/consoleFull) for   PR 4677 at commit [`b48a70f`](https://github.com/apache/spark/commit/b48a70fc168ddc673bda57208582d9af0f06d8b4).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27776 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27776/consoleFull) for   PR 4677 at commit [`b48a70f`](https://github.com/apache/spark/commit/b48a70fc168ddc673bda57208582d9af0f06d8b4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27776/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala:
```diff
@@ -76,8 +77,44 @@ class GradientBoostedTrees(private val boostingStrategy: BoostingStrategy)
+   * @param trainInput Training dataset: RDD of [[org.apache.spark.mllib.regression.LabeledPoint]].
+   * @param validateInput Validation dataset:
+                          RDD of [[org.apache.spark.mllib.regression.LabeledPoint]].
+                          Should follow same distribution as trainInput.
```
I think talking about target distributions may confuse some people.  Maybe we can clarify as follows:
```
This dataset should follow the same distribution as trainInput; e.g., these two datasets could be created from an original dataset by using [[org.apache.spark.rdd.RDD.randomSplit()]].
```

jkbradley replied:
Oh, and we should also explicitly say that validateInput should be a *different* dataset than trainInput.  (We don't need to check for this, though.  If it is the same, then validationTol acts like convergenceTol.)

jkbradley commented on docs/mllib-ensembles.md:
```diff
@@ -427,6 +427,18 @@ We omit some decision tree parameters since those are covered in the [decision t
 
+#### Validation while training
+
+Gradient boosting can overfit when trained with more number of trees. In order to prevent overfitting, it might
```
"with more number of trees" --> "with more trees"
"it might be" --> "it is"

jkbradley commented on docs/mllib-ensembles.md:
```diff
@@ -427,6 +427,18 @@ We omit some decision tree parameters since those are covered in the [decision t
+option. It takes a pair of RDD's as arguments, the first one being the training dataset and the second being the validation dataset.
+
+The training is stopped when the improvement in the validation error is not more than a certain tolerance
+(supplied by the **`validationTol`** argument in **`BoostingStrategy`**). In practice, the validation error
```
We generally don't use bold for arguments in the docs, but you could link to the API docs.

jkbradley commented on docs/mllib-ensembles.md:
```diff
@@ -427,6 +427,18 @@ We omit some decision tree parameters since those are covered in the [decision t
+
+The training is stopped when the improvement in the validation error is not more than a certain tolerance
+(supplied by the **`validationTol`** argument in **`BoostingStrategy`**). In practice, the validation error
+decreases with the increase in number of trees and then increases as the model starts to overfit. There might
```
"decreases with the increase in number of trees and then increases" --> "decreases initially and later increases"

jkbradley commented on docs/mllib-ensembles.md:
```diff
@@ -427,6 +427,18 @@ We omit some decision tree parameters since those are covered in the [decision t
+The training is stopped when the improvement in the validation error is not more than a certain tolerance
+(supplied by the **`validationTol`** argument in **`BoostingStrategy`**). In practice, the validation error
+decreases with the increase in number of trees and then increases as the model starts to overfit. There might
+be cases, in which the validation error does not change monotonically, and the user is advised to set a large
```
"be cases, in which" --> (no comma)

jkbradley commented on docs/mllib-ensembles.md:
```diff
@@ -427,6 +427,18 @@ We omit some decision tree parameters since those are covered in the [decision t
+(supplied by the **`validationTol`** argument in **`BoostingStrategy`**). In practice, the validation error
+decreases with the increase in number of trees and then increases as the model starts to overfit. There might
+be cases, in which the validation error does not change monotonically, and the user is advised to set a large
+enough negative tolerance and examine the validation curve to make further inference.
```
"to make further inference" --> "to tune the number of iterations"

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala:
```diff
@@ -177,6 +222,24 @@ object GradientBoostedTrees extends Logging {
+            boostingStrategy.treeStrategy.algo,
+            baseLearners.slice(0, bestM),
+            baseLearnerWeights.slice(0, bestM))
+        }
```
formatting:
```
} else if (stuff) {
```

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/tree/GradientBoostedTreesSuite.scala:
```diff
@@ -158,6 +158,63 @@ class GradientBoostedTreesSuite extends FunSuite with MLlibTestSparkContext {
+    val errorWithoutValidation = LogLoss.computeError(gbtRegressor, remappedInput)
+
+    assert(errorWithValidation < errorWithoutValidation)
+    }
```
fix indentation

jkbradley commented:
I still need to make a final pass, but it's looking good!

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala:
```diff
@@ -76,8 +77,44 @@ class GradientBoostedTrees(private val boostingStrategy: BoostingStrategy)
+        // Map labels to -1, +1 so binary classification can be treated as regression.
+        val remappedTrainInput = trainInput.map(
+          x => new LabeledPoint((x.label * 2) - 1, x.features))
+        val remappedValidateInput = trainInput.map(
```
"trainInput" --> "validateInput"

MechCoder replied:
oops. :/

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala:
```diff
@@ -108,12 +145,16 @@ object GradientBoostedTrees extends Logging {
-      boostingStrategy: BoostingStrategy): GradientBoostedTreesModel = {
+      validateInput: RDD[LabeledPoint],
+      boostingStrategy: BoostingStrategy,
+      validate: Boolean = false): GradientBoostedTreesModel = {
```
no need for default value; better to be explicit internally

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala:
```diff
@@ -177,6 +222,24 @@ object GradientBoostedTrees extends Logging {
+
+      if (validate) {
+        // Stop training early if
+        // 1. Reduction in error is lesser than the validationTol or
```
"lesser" --> "less"

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/configuration/BoostingStrategy.scala:
```diff
@@ -34,6 +34,12 @@ import org.apache.spark.mllib.tree.loss.{LogLoss, SquaredError, Loss}
  *                      weak hypotheses used in the final model.
  * @param learningRate Learning rate for shrinking the contribution of each estimator. The
  *                     learning rate should be between in the interval (0, 1]
+ * @param validationTol Useful when runWithValidation is used. If the error rate between two
```
Also: "the error rate" --> "the error rate on the validationInput"

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/configuration/BoostingStrategy.scala:
```diff
@@ -34,6 +34,12 @@ import org.apache.spark.mllib.tree.loss.{LogLoss, SquaredError, Loss}
  * @param learningRate Learning rate for shrinking the contribution of each estimator. The
  *                     learning rate should be between in the interval (0, 1]
+ * @param validationTol Useful when runWithValidation is used. If the error rate between two
+                        iterations is lesser than the validationTol, then stop. If run
```
"lesser" --> "less"

Also, use double brackets to add link to "run" to be clearer.

jkbradley commented:
Thinking more about naming parameters, I'd vote for this change in runWithValidation to be more consistent:
* "trainInput" --> "input" (consistent with run)
* "validateInput" --> "validationInput" (consistent with validationTol)

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/configuration/BoostingStrategy.scala:
```diff
@@ -34,6 +34,12 @@ import org.apache.spark.mllib.tree.loss.{LogLoss, SquaredError, Loss}
+ * @param validationTol Useful when runWithValidation is used. If the error rate between two
+                        iterations is lesser than the validationTol, then stop. If run
+                        is used, then this parameter is ignored.
+
```
Something weird is going on with the indentation and the lack of "*" (for comments) here

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/configuration/BoostingStrategy.scala:
```diff
@@ -34,6 +34,12 @@ import org.apache.spark.mllib.tree.loss.{LogLoss, SquaredError, Loss}
+                        is used, then this parameter is ignored.
+
+ a pair of RDD's are supplied to run. If the error rate
+ *                      between two iterations is lesser than convergenceTol, then training stops.
```
"lesser" --> "less"

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/tree/GradientBoostedTreesSuite.scala:
```diff
@@ -158,6 +158,63 @@ class GradientBoostedTreesSuite extends FunSuite with MLlibTestSparkContext {
+    // The error checked for internally in the GradientBoostedTrees is based on Regression.
+    // Hence for the validation model, the Classification error need not be strictly less than
+    // that done with validation.
+    val gbtValidateRegressor = new GradientBoostedTreesModel(
```
Hm, I misunderstood this the first time you asked about it.  It's weird to create a regression model and test using LogLoss.  I would test on validateRdd, not on trainRdd.  That's really all we need to check.  And it should let you keep the model a Classification model.

MechCoder replied:
I have addressed all your comment except this.
I am testing with validationInput only. Sorry if the variable name is confusing.

I think what happens is the number of true labels classified is the same whether or not I run with validation in because of the dataset that is being tested here. i.e when I run without validation, there might be an increase in the validation error but there is no change in the number of labels that are predicted correctly.

MechCoder replied:
and I'm not sure it's that weird, because that is what is being done internally :P , unless you have other ideas to test this.

jkbradley replied:
Oh, I

jkbradley replied:
Oh, I got confused about which dataset remappedInput was from.  In that case, I think it's just a flaky test.  I think it would be sufficient to check for error <= instead of <, especially since you are already checking that it stops early.

jkbradley commented:
Final comment: Would it shorten the code to combine the Classification and Regression tests?  You could loop over (algo, loss) pairs in a single loop.

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/tree/GradientBoostedTreesSuite.scala:
```diff
@@ -158,6 +158,63 @@ class GradientBoostedTreesSuite extends FunSuite with MLlibTestSparkContext {
   }
+
+  test("runWithValidation performs better on a validation dataset (Regression)") {
+    // Set numIterations large enough so that it early stops.
```
"early stops" --> "stops early"

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/tree/GradientBoostedTreesSuite.scala:
```diff
@@ -158,6 +158,63 @@ class GradientBoostedTreesSuite extends FunSuite with MLlibTestSparkContext {
+      val boostingStrategy =
+        new BoostingStrategy(treeStrategy, error, numIterations, validationTol = 0.0)
+
+      val gbtValidate = new GradientBoostedTrees(boostingStrategy).runWithValidation(
```
formatting: Put .runWithValidation on the next line with its parameters

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/tree/GradientBoostedTreesSuite.scala:
```diff
@@ -158,6 +158,63 @@ class GradientBoostedTreesSuite extends FunSuite with MLlibTestSparkContext {
+
+      val gbtValidate = new GradientBoostedTrees(boostingStrategy).runWithValidation(
+        trainRdd, validateRdd)
+      assert(gbtValidate.numTrees != numIterations)
```
Use ```!==``` (handles types better)

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/tree/GradientBoostedTreesSuite.scala:
```diff
@@ -158,6 +158,63 @@ class GradientBoostedTreesSuite extends FunSuite with MLlibTestSparkContext {
+  }
+
+  test("runWithValidation performs better on a validation dataset (Classification)") {
+    // Set numIterations large enough so that it early stops.
```
ditto: "stops early"

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/tree/GradientBoostedTreesSuite.scala:
```diff
@@ -158,6 +158,63 @@ class GradientBoostedTreesSuite extends FunSuite with MLlibTestSparkContext {
+      new BoostingStrategy(treeStrategy, LogLoss, numIterations, validationTol = 0.0)
+
+    // Test that it stops early.
+    val gbtValidate = new GradientBoostedTrees(boostingStrategy).runWithValidation(
```
move .runWithValidation to next line

jkbradley commented:
OK, I've made a close pass, so hopefully those are my final comments.

MechCoder committed:
Addresses indentation and doc comments

SparkQA commented:
[Test build #27847 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27847/consoleFull) for   PR 4677 at commit [`e4d799b`](https://github.com/apache/spark/commit/e4d799b679c519125a07851fc90a53244177a196).
 * This patch merges cleanly.

MechCoder commented:
@jkbradley Addressed all your comments except the inline one.

SparkQA commented:
[Test build #27847 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27847/consoleFull) for   PR 4677 at commit [`e4d799b`](https://github.com/apache/spark/commit/e4d799b679c519125a07851fc90a53244177a196).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27847/
Test PASSed.

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/tree/GradientBoostedTreesSuite.scala:
```diff
@@ -158,6 +158,63 @@ class GradientBoostedTreesSuite extends FunSuite with MLlibTestSparkContext {
+
+    // Test that it stops early.
+    val gbtValidate = new GradientBoostedTrees(boostingStrategy).
+      runWithValidation(trainRdd, validateRdd)
```
please put period (.) on line with runWithValidation:
```
    val gbtValidate = new GradientBoostedTrees(boostingStrategy)
      .runWithValidation(trainRdd, validateRdd)
```

jkbradley commented on mllib/src/test/scala/org/apache/spark/mllib/tree/GradientBoostedTreesSuite.scala:
```diff
@@ -158,6 +158,63 @@ class GradientBoostedTreesSuite extends FunSuite with MLlibTestSparkContext {
+        new BoostingStrategy(treeStrategy, error, numIterations, validationTol = 0.0)
+
+      val gbtValidate = new GradientBoostedTrees(boostingStrategy).
+        runWithValidation(trainRdd, validateRdd)
```
please put period (.) on line with runWithValidation:
```
    val gbtValidate = new GradientBoostedTrees(boostingStrategy)
      .runWithValidation(trainRdd, validateRdd)
```

jkbradley commented on docs/mllib-ensembles.md:
```diff
@@ -427,6 +427,17 @@ We omit some decision tree parameters since those are covered in the [decision t
+first one being the training dataset and the second being the validation dataset.
+
+The training is stopped when the improvement in the validation error is not more than a certain tolerance
+(supplied by the validationTol argument in BoostingStrategy). In practice, the validation error
```
I'd keep the backticks: ``` `validationTol` ``` (same for BoostingStrategy).  But the asterisks for bold are not needed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala:
```diff
@@ -177,6 +224,23 @@ object GradientBoostedTrees extends Logging {
+            boostingStrategy.treeStrategy.algo,
+            baseLearners.slice(0, bestM),
+            baseLearnerWeights.slice(0, bestM))
+        } else if (currentValidateError < bestValidateError){
```
scala style: space before {

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/tree/configuration/BoostingStrategy.scala:
```diff
@@ -34,7 +34,11 @@ import org.apache.spark.mllib.tree.loss.{LogLoss, SquaredError, Loss}
+ *                      validation input between two iterations is less than the validationTol
+ *                      then stop. Ignored when [[run]] is used.
  */
+
```
Remove empty line

jkbradley commented:
Also, what about shortening the code to combine the Classification and Regression tests? Did that not work out?

MechCoder committed:
Combine regression and classification tests into a single one

SparkQA commented:
[Test build #27889 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27889/consoleFull) for   PR 4677 at commit [`1bb21d4`](https://github.com/apache/spark/commit/1bb21d410f7bd59855b58cf606f36a43a19a7ee0).
 * This patch merges cleanly.

MechCoder commented:
@jkbradley I have fixed up your comments ! Hopefully good to go.

[off-topic]
It would be really great and helpful if Spark would be interested in taking students under Apache Software Foundation (https://community.apache.org/gsoc.html) for Google Summer of Code (https://www.google-melange.com/gsoc/homepage/google/gsoc2015) (assuming ASF gets selected of course), considering the fact that there is some interest right now. I had posted it in the developers list a few days ago, but it seems that all the developers are busy. Would you or @mengxr (or other MLlib developers) be interested in mentoring a MLlib related project this summer? If yes, then we could brainstorm on the JIRA issues on what could possibly be worked on this summer and I could start writing a proposal.
Thanks. :)

SparkQA commented:
[Test build #27889 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27889/consoleFull) for   PR 4677 at commit [`1bb21d4`](https://github.com/apache/spark/commit/1bb21d410f7bd59855b58cf606f36a43a19a7ee0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27889/
Test PASSed.

mengxr commented:
@MechCoder Thanks for ping us about GSoC! Please check my reply on the dev list.

jkbradley commented:
LGTM  I'll merge it into master

jkbradley commented:
Done---thanks for the PR!
